ouroboros-network-framework/src/Ouroboros/Network/InboundGovernor.hs

```diff
@@ -139,35 +156,165 @@ with :: forall (muxMode :: Mux.Mode) socket initiatorCtx peerAddr versionData ve
-      idleTimeout        = idleTimeout,
-      connectionManager  = connectionManager
+      transitionTracer = trTracer,
+      tracer,
+      debugTracer,
+      connectionDataFlow,
+      idleTimeout,
+      infoChannel,
+      withConnectionManager,
+      mkConnectionHandler
     }
     k
     = do
     labelThisThread "inbound-governor"
-    var <- newTVarIO (mkPublicState emptyState)
-    withAsync ((do
-               labelThisThread "inbound-governor-loop"
-               inboundGovernorLoop var emptyState)
-                `catch`
-               handleError var) $
+    stateVar <- newTVarIO emptyState
+    countersVar <- newTVarIO Map.empty
+    -- ^ warm and hot responder miniprotocol counters for a peer
+    -- to track and respond to transitions
+    let connectionHandler = mkConnectionHandler $ responderIgTracer stateVar countersVar
+    withConnectionManager connectionHandler \connectionManager ->
+      withAsync
+        (  labelThisThread "inbound-governor-loop" >>
+           forever (inboundGovernorStep connectionManager stateVar)
+         `catch`
+           handleError stateVar)
       \thread ->
-        k thread (readTVarIO var)
+        k thread (mkPublicState <$> readTVarIO stateVar) connectionManager
   where
+    -- the responder IG info channel tracer is embedded with the mux tracer unconditionally
+    -- by the conn handler on the inbound side and conditionally for 'InitiatorResponderMode'
+    -- on the outbound side only when the latter has negotiated a 'Duplex' data flow - otherwise
+    -- for unidirectional mode the IG is never provided with the peer handle since the inbound
+    -- is inactive.
+    responderIgTracer :: StrictTVar m (State muxMode initiatorCtx peerAddr versionData m a b)
+                      -> StrictTVar m (Map (ConnectionId peerAddr) ResponderCounters)
+                      -> Tracer m (Mux.WithBearer (ConnectionId peerAddr) Mux.Trace)
+    responderIgTracer stateVar countersVar = Tracer $ \(Mux.WithBearer peer trace) ->
+      -- hello from muxer main thread
+      case trace of
+        Mux.TraceState Mux.Mature -> atomically do
+          -- For inbound and outbound duplex connections, following a successful
+          -- handshake, the muxer begins racing with the connection
+          -- manager on informing the IG of the new connection
+          -- and responder miniprotocols' activity. In principle,
+          -- the problematic case is when the IG tracer registers responder
+          -- miniprotocol launch command before the CM informs the IG of the
+          -- new connection. However unlikely, this will lead to
+          -- incoherent transitions between the two components.
+          -- The muxer blocks itself here to yield right-of-way
+          -- to the CM.
+          check . Map.member peer . connections =<< readTVar stateVar
+          modifyTVar countersVar $ Map.insert peer (ResponderCounters 0 0)
+
+        (startWithID -> Just mid) -> atomically do
```

Comment:
I don't think you need view patterns here, hence can avoid adding a language extension.

If you do `_ | Just mid <- startWithID ->` I think it works.

Then you can better organise all your other clauses under the same match